### PR TITLE
Feature/issue13 non excel support

### PR DESCRIPTION
**Pull request by cschu**
*plink*

### CODE REVIEW

**Comment by MoritzHahn1337** [comment]
Tested with csv files using colon, comma, semicolon and tabs as separators

**Review by MoritzHahn1337 — APPROVED** [review]

**Review by MoritzHahn1337 — APPROVED** [review]

File: datameta/api/ui/convert.py [hunk 1]
```diff
@@ -36,12 +37,35 @@ def formatted_mrec_value(value, datetimefmt):
...
+        reader = make_excel_reader
+    else:
+        file_like_obj.seek(0)
+        dialect = csv.Sniffer().sniff(file_like_obj.read(1024).decode())
+        def make_table_reader(sheet, sep=dialect.delimiter):
+            sheet.seek(0)
+            return pd.read_table(sheet, dtype="object", sep=sep)
+        reader = make_table_reader
+
+    return reader
```
Comment (lkuchenb) [inline datameta/api/ui/convert.py:59]: I've changed this slightly to improve code readability. Also I think it's best to always change the file access position to `0` before returning it and not to rely on the fact that the caller will immediately use the returned func.

```suggestion
def get_samplesheet_reader(file_like_obj):
    """Given a file with tabular data which is either in delimited plain text
    format or in XLS(X) format, returns a function capable of reading the file
    and returning a pandas.DataFrame"""
    # https://readxl.tidyverse.org/reference/excel_format.html
    xlsx_sig   = b'PK\x03\x04', 
    xls_sig    = b'\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1'
    nibble     = file_like_obj.read(8)

    file_like_obj.seek(0)

    if nibble.startswith(xlsx_sig) or nibble.startswith(xls_sig):
        return lambda sheet : pd.read_excel(sheet, dtype="object")
    else:
        dialect = csv.Sniffer().sniff(file_like_obj.read(1024).decode())
        file_like_obj.seek(0)
        return lambda sheet : pd.read_table(sheet, dtype="object", sep=dialect.delimiter)
```

File: datameta/api/ui/convert.py [hunk 2]
```diff
@@ -36,12 +37,35 @@ def formatted_mrec_value(value, datetimefmt):
...
+
+
 ####################################################################################################
 
 def convert_samplesheet(db, file_like_obj, filename, user):
     # Try to read the sample sheet
     try:
-        submitted_metadata = pd.read_excel(file_like_obj, dtype="object")
+        reader = determine_samplesheet_reader(file_like_obj)
```
Comment (lkuchenb) [inline datameta/api/ui/convert.py:67]:
```suggestion
        reader = get_samplesheet_reader(file_like_obj)
```

**Review by lkuchenb — CHANGES_REQUESTED** [review]
Thank you, that seems to work nicely!

File: datameta/api/ui/convert.py (hunk 1 above)
Comment (cschu) [reply datameta/api/ui/convert.py:59]: @lkuchenb you mean lambdas are more readable than closures? :P

Comment (lkuchenb) [reply datameta/api/ui/convert.py:59]: Nothing I'd be religious about, but after moving the seek into the main function scope the functions became one-liners ;-)

Comment (cschu) [reply datameta/api/ui/convert.py]: Fair enough. I kept the closures as I think they're easier to read than the lambdas (prefer lambdas in map/filter context).

File: datameta/api/ui/convert.py [hunk 3]
```diff
@@ -36,12 +37,36 @@ def formatted_mrec_value(value, datetimefmt):
...
+            return pd.read_table(sheet, dtype="object", sep=sep)
+        file_like_obj.seek(0)
+        return create_table_reader
+
+
 ####################################################################################################
 
 def convert_samplesheet(db, file_like_obj, filename, user):
-    # Try to read the sample sheet
+    # Try to read the sample sheet
```
Comment (lkuchenb) [inline datameta/api/ui/convert.py:66]:
```suggestion
    # Try to read the sample sheet
```

**Review by lkuchenb — APPROVED** [review]
👍

File: datameta/api/ui/convert.py (hunk 1 above)
Comment (lkuchenb) [reply datameta/api/ui/convert.py]: Fine with me 😄

**Pull request merged**